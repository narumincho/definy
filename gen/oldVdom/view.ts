import { Language, TwitterCard } from "./data";

export type View<Message> = {
  /**
   * ページ名
   *
   * Google 検索のページ名や, タブ, ブックマークのタイトル, OGPのタイトルなどに使用される
   */
  readonly pageName: string;

  /**
   * アプリ名 / サイト名 (HTML出力のみ反映)
   */
  readonly appName: string;

  /**
   * ページの説明 (HTML出力のみ反映)
   */
  readonly description: string;

  /**
   * テーマカラー
   */
  readonly themeColor: Color | undefined;

  /**
   * アイコン画像のパス. /から始まる必要がある
   */
  readonly iconPath: string;

  /**
   * 使用している言語
   */
  readonly language: Language | undefined;

  /**
   * OGPに使われるカバー画像のURL (CORSの制限を受けない)
   */
  readonly coverImageUrl: URL;

  /** ページのURL */
  readonly url: URL;

  /** マニフェストのパス */
  readonly manifestPath?: ReadonlyArray<string>;

  /** Twitter Card。Twitterでシェアしたときの表示をどうするか */
  readonly twitterCard: TwitterCard;

  /** 全体に適応されるスタイル. CSS */
  readonly style?: string;

  /** スタイルのURL */
  readonly styleUrlList?: ReadonlyArray<URL>;

  /** ES Modules形式のJavaScript */
  readonly script?: string;

  /** メインのスクリプトのパス */
  readonly scriptPath?: string;

  /** スクリプトのURL. 違うオリジンのスクリプトを含めたいときに使う */
  readonly scriptUrlList?: ReadonlyArray<URL>;

  /** body の class */
  readonly bodyClass: string;

  /** 表示領域内でマウスや, タッチが動いたときのメッセージ */
  readonly pointerMove?: (pointer: Pointer) => Message;

  /** 表示領域内でマウスがクリックされたり, 画面に接触したときのメッセージ */
  readonly pointerDown?: (pointer: Pointer) => Message;
  /** body の 子要素 */
  readonly children: Children<Message>;
};

export type Pointer = {
  /** イベントの原因となっているポインタの一意の識別子 */
  pointerId: number;
  /** ポインタの接触ジオメトリの幅 */
  width: number;
  /** ポインタの接触ジオメトリの高さ */
  height: number;
  /** 0 から 1 の範囲のポインタ入力の正規化された圧力。 ここで、0 と 1 は、それぞれハードウェアが検出できる最小圧力と最大圧力を表します。 */
  pressure: number;
  /** ポインタ入力の正規化された接線圧力（バレル圧力またはシリンダー応力（cylinder stress）とも呼ばれます）は -1 から 1 の範囲で、0 はコントロールの中立位置です。 */
  tangentialPressure: number;
  /** Y-Z 平面と、ポインタ（ペン/スタイラスなど）軸と Y 軸の両方を含む平面との間の平面角度（度単位、-90 から 90 の範囲）。 */
  tiltX: number;
  /** X-Z 平面と、ポインタ（ペン/スタイラスなど）軸と X 軸の両方を含む平面との間の平面角度（度単位、-90 から 90 の範囲）。 */
  tiltY: number;
  /** ポインタ（ペン/スタイラスなど）の長軸を中心とした時計回りの回転の度数（0 から 359の範囲の値）。 */
  twist: number;
  /** イベントの原因となったデバイスタイプ（マウス、ペン、タッチなど）を示します。 */
  pointerType: PointerType;
  /** ポインタがこのポインタタイプのプライマリポインタを表すかどうかを示します。 */
  isPrimary: boolean;
  /** 表示領域のX座標 */
  x: number;
  /** 表示領域のY座標 */
  y: number;
};

/** ポインターの種類 */
export type PointerType = "mouse" | "pen" | "touch" | "";

/** メッセージを集計した結果 */
export type MessageData<Message> = {
  readonly messageMap: ReadonlyMap<string, Events<Message>>;
  readonly pointerMove: ((pointer: Pointer) => Message) | undefined;
  readonly pointerDown: ((pointer: Pointer) => Message) | undefined;
};
/**
 * View の 差分データ.
 * イベント関係は差分を使って処理をしないので Message は含まれないが, 要素を追加するときに Message を使う形になってしまっている
 */
export type ViewDiff<Message> = {
  readonly patchOperationList: ReadonlyArray<ViewPatchOperation>;
  readonly childrenDiff: ChildrenDiff<Message>;
  readonly newMessageData: MessageData<Message>;
};

export type ViewPatchOperation =
  | { tag: "changePageName"; newTitle: string }
  | { tag: "changeThemeColor"; newThemeColor: Color | undefined }
  | { tag: "changeLanguage"; newLanguage: Language | undefined }
  | { tag: "changeBodyClass"; newClass: string };

export type Element<Message> =
  | Div<Message>
  | ExternalLink<Message>
  | LocalLink<Message>
  | Button<Message>
  | Img
  | InputRadio<Message>
  | InputText<Message>
  | TextArea<Message>
  | Label<Message>
  | Svg<Message>
  | SvgPath
  | SvgCircle
  | SvgAnimate;

export type ElementDiff<Message> =
  | {
      readonly kind: "replace";
      readonly newElement: Element<Message>;
      readonly key: string;
    }
  | {
      readonly kind: "update";
      readonly elementUpdateDiff: ElementUpdateDiff<Message>;
      readonly key: string;
    }
  | {
      readonly kind: "delete";
    }
  | {
      readonly kind: "insert";
      readonly element: Element<Message>;
      readonly key: string;
    }
  | {
      readonly kind: "skip";
    };

export type ElementUpdateDiff<Message> =
  | DivDiff<Message>
  | ExternalLinkDiff<Message>
  | LocalLinkDiff<Message>
  | ButtonDiff<Message>
  | ImgDiff
  | InputRadioDiff
  | InputTextDiff
  | TextAreaDiff
  | LabelDiff<Message>
  | SvgDiff<Message>
  | SvgPathDiff
  | SvgCircleDiff
  | SvgAnimateDiff;

export type Div<Message> = {
  readonly tag: "div";
  readonly id: string;
  readonly class: string;
  readonly click: ClickMessageData<Message> | null;
  readonly children: Children<Message>;
};

export type DivDiff<Message> = {
  readonly tag: "div";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly children: ChildrenDiff<Message>;
};

export type ExternalLink<Message> = {
  readonly tag: "externalLink";
  readonly id: string;
  readonly class: string;
  readonly url: string;
  readonly children: Children<Message>;
};

export type ExternalLinkDiff<Message> = {
  readonly tag: "externalLink";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly url: string | undefined;
  readonly children: ChildrenDiff<Message>;
};

export type LocalLink<Message> = {
  readonly tag: "localLink";
  readonly id: string;
  readonly class: string;
  readonly url: string;
  readonly jumpMessage: Message;
  readonly children: Children<Message>;
};

export type LocalLinkDiff<Message> = {
  readonly tag: "localLink";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly url: string | undefined;
  readonly children: ChildrenDiff<Message>;
};

export type Button<Message> = {
  readonly tag: "button";
  readonly id: string;
  readonly class: string;
  readonly click: Message;
  readonly children: Children<Message>;
};

export type ButtonDiff<Message> = {
  readonly tag: "button";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly children: ChildrenDiff<Message>;
};

export type Img = {
  readonly tag: "img";
  readonly id: string;
  readonly class: string;
  readonly alt: string;
  readonly src: string;
};

export type ImgDiff = {
  readonly tag: "img";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly alt: string | undefined;
  readonly src: string | undefined;
};

export type InputRadio<Message> = {
  readonly tag: "inputRadio";
  readonly id: string;
  readonly class: string;
  readonly select: Message;
  readonly checked: boolean;
  /** 選択肢の選択を1にする動作のため. どの選択肢に属しているかを指定する */
  readonly name: string;
};

export type InputRadioDiff = {
  readonly tag: "inputRadio";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly checked: boolean | undefined;
  readonly name: string | undefined;
};

export type InputText<Message> = {
  readonly tag: "inputText";
  readonly id: string;
  readonly class: string;
  readonly inputOrReadonly: ((text: string) => Message) | null;
  readonly value: string;
};

export type InputTextDiff = {
  readonly tag: "inputText";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly readonly: boolean | undefined;
  readonly value: string | undefined;
};

export type TextArea<Message> = {
  readonly tag: "textArea";
  readonly id: string;
  readonly class: string;
  readonly inputOrReadonly: ((text: string) => Message) | null;
  readonly value: string;
};

export type TextAreaDiff = {
  readonly tag: "textArea";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly readonly: boolean | undefined;
  readonly value: string | undefined;
};

export type Label<Message> = {
  readonly tag: "label";
  readonly id: string;
  readonly class: string;
  readonly for: string;
  readonly children: Children<Message>;
};

export type LabelDiff<Message> = {
  readonly tag: "label";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly for: string | undefined;
  readonly children: ChildrenDiff<Message>;
};

export type Svg<Message> = {
  readonly tag: "svg";
  readonly id: string;
  readonly class: string;
  readonly viewBoxX: number;
  readonly viewBoxY: number;
  readonly viewBoxWidth: number;
  readonly viewBoxHeight: number;
  readonly children: Children<Message>;
};

export type SvgDiff<Message> = {
  readonly tag: "svg";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly viewBoxX: number | undefined;
  readonly viewBoxY: number | undefined;
  readonly viewBoxWidth: number | undefined;
  readonly viewBoxHeight: number | undefined;
  readonly children: ChildrenDiff<Message>;
};

export type SvgPath = {
  readonly tag: "path";
  readonly id: string;
  readonly class: string;
  readonly d: string;
  readonly fill: string;
};

export type SvgPathDiff = {
  readonly tag: "path";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly d: string | undefined;
  readonly fill: string | undefined;
};

export type SvgCircle = {
  readonly tag: "circle";
  readonly id: string;
  readonly class: string;
  readonly fill: string;
  readonly stroke: string;
  readonly cx: number;
  readonly cy: number;
  readonly r: number;
  readonly children: Children<never>;
};

export type SvgCircleDiff = {
  readonly tag: "circle";
  readonly id: string | undefined;
  readonly class: string | undefined;
  readonly fill: string | undefined;
  readonly stroke: string | undefined;
  readonly cx: number | undefined;
  readonly cy: number | undefined;
  readonly r: number | undefined;
  readonly children: ChildrenDiff<never>;
};

export type SvgAnimate = {
  readonly tag: "animate";
  readonly attributeName: string;
  readonly dur: number;
  readonly repeatCount: string;
  readonly from: string;
  readonly to: string;
};

export type SvgAnimateDiff = {
  readonly tag: "animate";
  readonly attributeName: string | undefined;
  readonly dur: number | undefined;
  readonly repeatCount: string | undefined;
  readonly from: string | undefined;
  readonly to: string | undefined;
};

/**
 * 各要素のイベントのハンドルをどうするかのデータ
 *
 * Message を undefined にするとイベントが設定されてないと解釈される場合があるので注意!
 */
export type Events<Message> = {
  onClick: ClickMessageData<Message> | undefined;
  onChange: ChangeMessageData<Message> | undefined;
  onInput: InputMessageData<Message> | undefined;
};

export type Path = string & { readonly _path: undefined };

export const rootPath = "" as Path;

export const pathAppendKey = (path: Path, key: string): Path =>
  (path + "/" + key) as Path;

export type ClickMessageData<Message> = {
  ignoreNewTab: boolean;
  stopPropagation: boolean;
  message: Message;
};

export type ChangeMessageData<Message> = Message;

export type InputMessageData<Message> = (value: string) => Message;

export const childrenElementListTag = Symbol("Children - ElementList");
export const childrenTextTag = Symbol("Children - Text");

export const childrenElementList = <Message>(
  value: ReadonlyMap<string, Element<Message>>
): Children<Message> => ({ tag: childrenElementListTag, value });

export const childrenText = <Message>(value: string): Children<Message> => ({
  tag: childrenTextTag,
  value,
});

export type Children<Message> =
  | {
      readonly tag: typeof childrenElementListTag;
      readonly value: ReadonlyMap<string, Element<Message>>;
    }
  | {
      readonly tag: typeof childrenTextTag;
      readonly value: string;
    };

export type ChildrenDiff<Message> =
  | {
      readonly kind: "skip";
    }
  | {
      readonly kind: "setText";
      readonly text: string;
    }
  | {
      readonly kind: "resetAndInsert";
      readonly value: ReadonlyMap<string, Element<Message>>;
    }
  | {
      readonly kind: "childDiffList";
      readonly children: ReadonlyArray<ElementDiff<Message>>;
    };

/** 色を表現する rgbは 0...1 の範囲でなければならない */
export type Color = {
  readonly r: number;
  readonly g: number;
  readonly b: number;
};
