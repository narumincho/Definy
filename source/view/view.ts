import * as d from "definy-core/source/data";

export interface View<Message> {
  readonly title: string;
  readonly themeColor: Color | undefined;
  readonly language: d.Language;
  readonly attributeAndChildren: AttributesAndChildren<Message>;
}

export interface Element<Message> {
  readonly tagName: string;
  readonly attributeAndChildren: AttributesAndChildren<Message>;
  readonly isSvg: boolean;
}

export interface AttributesAndChildren<Message> {
  readonly attributes: ReadonlyMap<string, string>;
  readonly events: Events<Message>;
  readonly children: Children<Message>;
}

/**
 * 各要素のイベントのハンドルをどうするかのデータ
 *
 * Message を undefined にするとイベントが設定されてないと解釈される場合があるので注意!
 */
export interface Events<Message> {
  onClick: ClickMessageData<Message> | undefined;
  onChange: ChangeMessageData<Message> | undefined;
  onInput: InputMessageData<Message> | undefined;
}

export type Path = string & { readonly _path: undefined };

export const rootPath = "" as Path;

export const pathAppendKey = (path: Path, key: string): Path =>
  (path + "/" + key) as Path;

export interface ClickMessageData<Message> {
  ignoreNewTab: boolean;
  message: Message;
}

export type ChangeMessageData<Message> = Message;

export type InputMessageData<Message> = (value: string) => Message;

export const childrenElementListTag = Symbol("Children - ElementList");
export const childrenTextTag = Symbol("Children - Text");

export const childrenElementList = <Message>(
  value: ReadonlyMap<string, Element<Message>>
): Children<Message> => ({ tag: childrenElementListTag, value });

export const childrenText = <Message>(value: string): Children<Message> => ({
  tag: childrenTextTag,
  value,
});

export type Children<Message> =
  | {
      readonly tag: typeof childrenElementListTag;
      readonly value: ReadonlyMap<string, Element<Message>>;
    }
  | {
      readonly tag: typeof childrenTextTag;
      readonly value: string;
    };

/** 色を表現する rgbは 0...1 の範囲でなければならない */
export interface Color {
  readonly r: number;
  readonly g: number;
  readonly b: number;
}
