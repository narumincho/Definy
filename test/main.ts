import {
  Children,
  Element,
  View,
  childrenElementList,
  childrenText,
} from "../source/view/view";
import {
  createChildrenDiff,
  createElementDiff,
  createViewDiff,
} from "../source/view/diff";
import { createPatchState, domToView, patchView } from "../source/view/patch";
import { div, view } from "../source/view/viewUtil";
import { OrderedMap } from "immutable";

describe("test", () => {
  it("same text return skip", () => {
    const children: Children<never> = childrenText("それな");
    expect(createChildrenDiff(children, children)).toMatchSnapshot();
  });
  it("different text return setText", () => {
    const oldChildren: Children<never> = childrenText("A");
    const newChildren: Children<never> = childrenText("B");
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });
  it("element to test return setText", () => {
    const oldChildren: Children<never> = childrenElementList(
      OrderedMap({ element: div<never>({}, "aaa") })
    );
    const newChildren: Children<never> = childrenText("new");
    expect(createChildrenDiff(oldChildren, newChildren));
  });
  it("add attribute", () => {
    const oldElement: Element<never> = div({}, "");
    const newElement: Element<never> = div({ id: "newId" }, "");
    expect(createElementDiff(oldElement, newElement)).toMatchSnapshot();
  });
  it("delete attribute", () => {
    const oldElement: Element<never> = div({ id: "newId" }, "");
    const newElement: Element<never> = div({}, "");
    expect(createElementDiff(oldElement, newElement)).toMatchSnapshot();
  });
  it("add element", () => {
    const oldChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        b: div<never>({}, "B"),
      })
    );
    const newChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        b: div<never>({}, "B"),
        c: div<never>({}, "newC"),
      })
    );
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });
  it("delete element", () => {
    const oldChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        b: div<never>({}, "B"),
        c: div<never>({}, "C"),
      })
    );
    const newChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        c: div<never>({}, "C"),
      })
    );
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });
  it("replace element", () => {
    const oldChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        b: div<never>({}, "B"),
        c: div<never>({}, "C"),
      })
    );
    const newChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        new: div<never>({}, "new"),
        c: div<never>({}, "C"),
      })
    );
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });
  it("insert element", () => {
    const oldChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        b: div<never>({}, "B"),
      })
    );
    const newChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        new: div<never>({}, "new"),
        b: div<never>({}, "B"),
      })
    );
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });
  it("complex diff", () => {
    const oldChildren: Children<never> = childrenElementList<never>(
      OrderedMap<Element<never>>({
        A: div({ id: "a" }, OrderedMap({ text: div<never>({}, "a") })),
        B: div({}, "B"),
        C: div(
          {},
          OrderedMap<Element<never>>({
            text: div({}, "C"),
            sub: div({}, "C sub"),
          })
        ),
        D: div({}, "D"),
        E: div({}, "E"),
        F: div({}, "F"),
      })
    );
    const newChildren: Children<never> = childrenElementList(
      OrderedMap<Element<never>>({
        A: div<never>({}, "A"),
        B: div({}, "B"),
        K: div({}, "K"),
        C: div(
          {},
          OrderedMap<Element<never>>({ text: div({}, "C") })
        ),
        S: div({}, "S"),
        E: div({}, "E"),
      })
    );
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });
  it("replace all", () => {
    const oldChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        b: div<never>({}, "B"),
        c: div<never>({}, "C"),
      })
    );
    const newChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        x: div<never>({}, "X"),
        y: div<never>({}, "Y"),
        z: div<never>({}, "Z"),
      })
    );
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });

  it("delete all children", () => {
    const oldChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        b: div<never>({}, "B"),
        c: div<never>({}, "C"),
        nest: div<never>(
          {},
          OrderedMap<Element<never>>({ inner: div<never>({}, "Inner") })
        ),
      })
    );
    const newChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        x: div<never>({}, "X"),
        y: div<never>({}, "Y"),
        z: div<never>({}, "Z"),
      })
    );
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });
  it("change key order", () => {
    const oldChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        a: div<never>({}, "A"),
        b: div<never>({}, "B"),
        c: div<never>({}, "C"),
      })
    );
    const newChildren: Children<never> = childrenElementList<never>(
      OrderedMap({
        b: div<never>({}, "B"),
        c: div<never>({}, "C"),
        a: div<never>({ id: "A" }, "A"),
      })
    );
    expect(createChildrenDiff(oldChildren, newChildren)).toMatchSnapshot();
  });
  it("use fake", () => {
    const oldView: View<never> = view<never>(
      {
        language: "Japanese",
        themeColor: { r: 0, g: 0, b: 0 },
        title: "",
      },
      ""
    );
    const newView: View<never> = view<never>(
      {
        language: "English",
        themeColor: { r: 0, g: 0, b: 0 },
        title: "",
      },
      OrderedMap({ target: div<never>({ id: "target" }, "new") })
    );
    const pathState = createPatchState(() => {});
    const diff = createViewDiff(oldView, newView);
    console.log(diff);
    patchView(diff, pathState);
    const domView = domToView();
    console.log(domView);
    if (domView._ === "Error") {
      throw new Error("DOMからViewを導出できなかった");
    }
    expect(domView.ok.attributeAndChildren).toEqual(
      newView.attributeAndChildren
    );
  });
});
