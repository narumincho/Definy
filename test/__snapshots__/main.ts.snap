// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test add attribute 1`] = `
Object {
  "attributeAndChildren": Object {
    "attributes": Object {
      "deleteNameSet": Immutable.Set [],
      "setNameValueMap": Immutable.Map {
        "id": "newId",
      },
    },
    "children": Object {
      "kind": "skip",
    },
    "events": Object {
      "deleteNameSet": Immutable.Set [],
      "setNameValueMap": Immutable.Map {},
    },
  },
  "kind": "update",
}
`;

exports[`test add element 1`] = `
Object {
  "children": Array [
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "newC",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "c",
      "kind": "insert",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test change key order 1`] = `
Object {
  "children": Array [
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "kind": "replace",
      "newElement": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "id": "A",
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "A",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test complex diff 1`] = `
Object {
  "children": Array [
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [
            "id",
          ],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "setText",
          "text": "A",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "K",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "K",
      "kind": "insert",
    },
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "children": Array [
            Object {
              "attributeAndChildren": Object {
                "attributes": Object {
                  "deleteNameSet": Immutable.Set [],
                  "setNameValueMap": Immutable.Map {},
                },
                "children": Object {
                  "kind": "skip",
                },
                "events": Object {
                  "deleteNameSet": Immutable.Set [],
                  "setNameValueMap": Immutable.Map {},
                },
              },
              "kind": "update",
            },
            Object {
              "kind": "delete",
            },
          ],
          "kind": "childDiffList",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "S",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "S",
      "kind": "insert",
    },
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "kind": "delete",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test delete all children 1`] = `
Object {
  "children": Array [
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "X",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "x",
      "kind": "insert",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "Y",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "y",
      "kind": "insert",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "Z",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "z",
      "kind": "insert",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test delete attribute 1`] = `
Object {
  "attributeAndChildren": Object {
    "attributes": Object {
      "deleteNameSet": Immutable.Set [
        "id",
      ],
      "setNameValueMap": Immutable.Map {},
    },
    "children": Object {
      "kind": "skip",
    },
    "events": Object {
      "deleteNameSet": Immutable.Set [],
      "setNameValueMap": Immutable.Map {},
    },
  },
  "kind": "update",
}
`;

exports[`test delete element 1`] = `
Object {
  "children": Array [
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test different text return setText 1`] = `
Object {
  "kind": "setText",
  "text": "B",
}
`;

exports[`test insert element 1`] = `
Object {
  "children": Array [
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "new",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "new",
      "kind": "insert",
    },
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test replace all 1`] = `
Object {
  "children": Array [
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "X",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "x",
      "kind": "insert",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "Y",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "y",
      "kind": "insert",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "Z",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "z",
      "kind": "insert",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test replace element 1`] = `
Object {
  "children": Array [
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "element": Object {
        "attributeAndChildren": Object {
          "attributes": Immutable.Map {
            "class": "css-0",
          },
          "children": Object {
            "tag": Symbol(Children - Text),
            "value": "new",
          },
          "events": Immutable.Map {},
        },
        "tagName": "div",
      },
      "key": "new",
      "kind": "insert",
    },
    Object {
      "attributeAndChildren": Object {
        "attributes": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
        "children": Object {
          "kind": "skip",
        },
        "events": Object {
          "deleteNameSet": Immutable.Set [],
          "setNameValueMap": Immutable.Map {},
        },
      },
      "kind": "update",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test same text return skip 1`] = `
Object {
  "kind": "skip",
}
`;
